### ISSUE TITLE
Text immediately following the line "--text follows this line--"

### ISSUE THREAD

**Issue by cuprum** [ahyatt/ekg#43]
It will be easier to show in a screenshot.

![1](https://user-images.githubusercontent.com/532546/235743280-202b69ab-0a63-4e36-9a48-9e0094249a20.gif)

That is, the text immediately following the line `--text follows this line--` is written to the metadata.

**Comment by ahyatt** [comment]
This probably I know about.  Really, the text is before the line, but the real problem is that text just after the line (the immediate character after the line) winds up going before the line instead of after.  It's weird, and I was already planning to look into it.  I think the best I can do is to move the text to after the line after the modification finishes.  It will probably remain weird.

Ideally, I'd like it to be impossible to even go to the line below the "text follows this line", but I don't think this is possible.

**Comment by ahyatt** [comment]
My fix does fix this issue (but please try it out), but introduces another issue that some commands like org-insert-heading don't work on an empty buffer.  However, I think many org changes like this seem problematic, since they assume they are in a normal buffer, which this is not.  I'll see if anything can be done about that, but for the moment I'm willing to live with this tradeoff.

**Comment by cuprum** [comment]
Maybe I found a solution that allows to get rid of these pitfalls.
Also, I commented out the hooks with `ekg--metadata-modification` thinking that function needs modification. I used text properties instead of overlay.

```emacs-lisp
(defun blank-line-motion-detect (window prevpos motion)
  (cond
   ((and (eq motion 'entered) (< prevpos (point)))
    (forward-line))
   ((and (eq motion 'entered) (> prevpos (point)))
    (forward-line -1))))

(defun ekg-edit-display-metadata ()
  "Create or edit the overlay to show metadata."
  (let ((o (ekg--metadata-overlay))
        (inhibit-read-only t))
    (buffer-disable-undo)
    (replace-region-contents (overlay-start o) (overlay-end o)
                             #'ekg--replace-metadata)
    (goto-char (overlay-end o))
    ;; (insert "\n")
    (move-overlay o (point-min) (- (overlay-end o) 1))
    ;; (overlay-put o 'after-string (propertize "--text follows this line--\n" 'read-only t))
    (insert (propertize "--text follows this line--\n"
			'read-only t
			'rear-nonsticky t ; allows press RET immediately after the line
			'cursor-sensor-functions '(blank-line-motion-detect)))
    (cursor-sensor-mode)
    (overlay-put o 'category 'ekg-metadata)
    ;; (overlay-put o 'modification-hooks '(ekg--metadata-modification))
    ;; (overlay-put o 'insert-behind-hooks '(ekg--metadata-modification))
    (overlay-put o 'face 'ekg-metadata)
    (buffer-enable-undo)
    ;; If org-mode is on, the metadata messes up the org-element-cache, so let's disable it.
    (when (eq major-mode 'org-mode)
      (setq-local org-element-use-cache nil))))
```

**Comment by qingshuizheng** [comment]
@cuprum 

> ```emacs-lisp
>     (insert (propertize "--text follows this line--\n"
> 			'read-only t
> 			'rear-nonsticky t ; allows press RET immediately after the line
> 			'cursor-sensor-functions '(blank-line-motion-detect)))
>     (cursor-sensor-mode)
> ```

`(insert (propertize ... ... ...))` unfortunately brings one side effect:  "--text follows this line--\n" will be counted as content text when save. Therefore, additional trim before save is expected. 



---
@ahyatt 

Another solution I use before 808921a was implemented, is to simply block myself from inserting any text immediately after "--text follows this line--\n".  (Note the second line from the bottom with "; <--- comment". )

```
(defun ekg--replace-metadata ()
  "Replace the metadata in a buffer."
  (let ((note ekg-note))
    (with-temp-buffer
      (when (ekg-should-show-id-p (ekg-note-id note))
        (insert (ekg--metadata-string "Resource" (ekg-note-id note))))
      (insert
       (ekg--metadata-string "Tags"
                             (mapconcat (lambda (tag) (format "%s" tag))
                                        (ekg-note-tags note) ", ")))
      (map-apply (lambda (k v)
                   (when-let (label (cdr (assoc k ekg-metadata-labels)))
                     (insert (ekg--metadata-string
                              label
                              (if (listp v)
                                  (mapconcat (lambda (v) (format "%s" v))
                                             v ", ")
                                (format "%s" v))))))
                 (ekg-note-properties note))
      (put-text-property (1- (point)) (point) 'read-only t)   ; <--- block user from inserting here
      (buffer-string))))
  
```

**Comment by cuprum** [comment]
@qingshuizheng Indeed, the line with the read-only property will have to be excluded.

**Comment by ahyatt** [comment]
Thanks, all.  I noticed a fairly bad bug with the `ekg--metadata-modification` causing problems, which I've fixed. I'll look into these alternate methods you mention, which look interesting.  Thank you for suggesting them!

**Comment by ahyatt** [comment]
I took a look, and the solution does work - but you still need metadata on-modify behavior, so the user can't insert a lot of blank lines or delete the metadata entirely.  I've checked in another simplification, not based on using text-properties at the moment.  I'm still thinking about this - but at the moment, I think the current solution is reasonable.  If it becomes intractable, both of these ideas are promising ways out.
